gcp/inference/handler.py

```diff
@@ -94,63 +105,69 @@ def preprocess(
+        tags = Prediction.generate_tags(open(local_path, "rb"), close=True)
+        img = cv2.cvtColor(cv2.imread(local_path), cv2.COLOR_BGR2RGB)
         Path(local_path).unlink()
 
-        # PREPROCESS IMAGE
-        s = img.shape[:2]
-        g = self.size / max(s)
-        shape1 = [[y * g for y in s]]
-        shape1 = [
-            make_divisible(x, self.max_stride) for x in np.stack(shape1, 0).max(0)
-        ]  # inference shape
-        x = letterbox(img, new_shape=shape1, auto=False)[0]  # pad
-        x = np.ascontiguousarray(x[None].transpose((0, 3, 1, 2)))  # BHWC to BCHW
-        img_tensor = (
-            torch.from_numpy(x).to(self.p.device).type_as(self.p) / 255.0
-        )  # uint8 to fp16/32
+        shape1, img_tensor = self.preprocess_img(img)
 
         print("HANDLER: Completed preprocessing")
-        return uri, img, img_tensor, shape1
+        return uri, img, img_tensor, shape1, tags
 
-    def inference(
-        self, data, *args, **kwargs
-    ) -> Tuple[str, np.ndarray, torch.Tensor, List, torch.Tensor]:
+    def inference(self, data, *args, **kwargs) -> Tuple[str, np.ndarray, List, Dict]:
         print("HANDLER: Starting inference")
-        uri, img, img_tensor, shape1 = data
+        uri, img, img_tensor, shape1, tags = data
         y = self.model(img_tensor)[0]
-        print("HANDLER: Completed inference")
-        return uri, img, img_tensor, shape1, y
-
-    def postprocess(self, data, *args, **kwargs):
-        print("HANDLER: Starting postprocessing")
-        uri, img, img_tensor, shape1, y = data
-
         y = non_max_suppression(
             y, self.conf, iou_thres=self.iou, classes=self.classes, max_det=self.max_det
-        )  # NMS
+        )
         scale_coords(shape1, y[0][:, :4], img.shape[:2])
         detections = Detections(
-            [img], y, ["test.jpg"], [0, 1, 2, 3], self.names, img_tensor.shape
+            imgs=[img], pred=y, files=[], times=[0, 1, 2, 3], names=CROP_CLASSES, shape=img_tensor.shape
         )
+        results = detections.pandas().xyxy[0].to_dict(orient="records")
 
-        uri_as_path = Path(uri)
+        print("HANDLER: Completed inference")
+        return uri, img, results, tags
 
-        local_dest_path = Path(
-            tempfile.gettempdir() + f"/result_{uri_as_path.stem}.json"
-        )
-        detections.pandas().xyxy[0].to_json(
-            path_or_buf=str(local_dest_path), indent=4, orient="records"
+    def postprocess(self, data, *args, **kwargs):
+        print("HANDLER: Starting postprocessing")
+        uri, img, results, tags = data
+        uri_as_path = Path(uri)
+        name = "-".join(uri_as_path.parts[2:-1]) + "-" + uri_as_path.stem
+        pred = Prediction.from_results_and_tags(
+            results=results, tags=tags, name=name, img_np=img
         )
+        save_to_db = pred.to_dict()
+        save_to_db["input_img"] = uri
+
+        labeled_uri = None
```

Comment:
If `LABEL_IMG_PERCENT=0.01` then 1% of all images will have their labeled image (with bounding boxes) uploaded to a bucket. 

If we upload a labeled image for every single input image that would be a big waste of time and space on the cloud.